code/modules/vueui/ui.dm

```diff
@@ -0,0 +1,321 @@
+  * Opens this ui, gathers initial data
+  *
+  * @return nothing
+  */
+/datum/vueui/proc/open()
+    if(QDELETED(object))
+        return
+    if(!user.client)
+        return
+    
+    if(!data)
+        data = object.vueui_data_change(null, user, src)
+    update_status()
+    if(!status || status == STATUS_CLOSE)
+        return
+
+    var/params = "window=[windowid];"
+    if(width && height)
+        params += "size=[width]x[height];"
+    send_resources_and_assets(user.client)
+    user << browse(generate_html(), params)
+    winset(user, "mapwindow.map", "focus=true")
+    spawn(1)
+        winset(user, windowid, "on-close=\"vueuiclose \ref[src]\"")
+
+/**
+  * Closes this ui
+  *
+  * @return nothing
+  */
+/datum/vueui/proc/close()
+    SSvueui.ui_closed(src)
+    user << browse(null, "window=[windowid]")
+    status = null
+    
+/**
+  * Generates base html for this ui to be rendered.
+  *
+  * @return html code - text
+  */
+/datum/vueui/proc/generate_html()
+#if UIDEBUG
+    var/debugtxt = "<div id=\"dapp\"></div>"
+#else
+    var/debugtxt = ""
+#endif
+    return {"
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta http-equiv="X-UA-Compatible" content="IE=11">
+        <meta charset="UTF-8">
+        <link rel="stylesheet" type="text/css" href="vueui.css">
+    </head>
+    <body class="[get_theme_class()]">
+        <div id="header">
+            <header-[header]></header-[header]>
+        </div>
+        <div id="app">
+
+        </div>
+        [debugtxt]
+        <noscript>
+            <div id='uiNoScript'>
+                <h2>JAVASCRIPT REQUIRED</h2>
+                <p>Your Internet Explorer's Javascript is disabled (or broken).<br/>
+                Enable Javascript and then open this UI again.</p>
+            </div>
+        </noscript>
+    </body>
+    <script type="application/json" id="initialstate">
+        [generate_data_json()]
+    </script>
+    <script type="text/javascript" src="vueui.js"></script>
+</html>
+    "}
+
+/**
+  * Generates json state object to be sent to ui.
+  *
+  * @return json object - text
+  */
+/datum/vueui/proc/generate_data_json()
+    var/list/sdata = list()
+    sdata["state"] = src.data
+    sdata["assets"] = list()
+    sdata["active"] = activeui
+    sdata["uiref"] = "\ref[src]"
+    sdata["status"] = status
+    sdata["title"] = title
+    sdata["wtime"] = world.time
+    for(var/asset_name in assets)
+        var/asset = assets[asset_name]
+        sdata["assets"][asset_name] = list("ref" = ckey("\ref[asset["img"]]"))
+    return json_encode(sdata)
+
+/**
+  * Sends all resources required for proper renderig of ui
+  *
+  * @param cl - client that should get assets
+  *
+  * @return nothing
+  */ 
+/datum/vueui/proc/send_resources_and_assets(var/client/cl)
+#if UIDEBUG
+    cl << browse_rsc(file("vueui/dist/main.js"), "vueui.js")
+    cl << browse_rsc(file("vueui/dist/main.css"), "vueui.css")
+#else
+    var/datum/asset/assets = get_asset_datum(/datum/asset/simple/vueui)
```

Comment:
I think there's a way to check if an asset has already been sent to the client instead of re-sending it each time?